### ISSUE TITLE
Inconsistent output discovery and enabling

### ISSUE THREAD

**Issue by ReeceStevens** [swaywm/sway#5843]
### Sway Verison
**Sway Version: 1.5.1**

### Logs 
I have uploaded the following relevant files [in this gist](https://gist.github.com/ReeceStevens/fbf2f213a7baa3f070e158e185e0e318):
- my sway config (it is not a minimal config, but I am fairly confident nothing in my configuration is contributing to this bug report)
- debug sway and DRM logs when multihead is working as expected
- debug sway and DRM logs when multihead is *not* working as expected. 

Please note that in the DRM logs, there is an error message that shows up quite frequently:

```
[107270.331733] msr: Write to unrecognized MSR 0x1a2 by python3
                Please report to x86@kernel.org
```

This is due to the use of a thinkpad thermal throttling workaround (bug details here: https://bugs.archlinux.org/task/68321), and I believe it is unrelated to the investigation here. However, it does produce a good bit of log noise. If you would like, I can filter these messages out and re-post the dmesg logs.

### Stack Trace
No sway crash, so no stack trace needed.

### Description

Firstly: thank you for the great work you all do to build out and maintain sway! I have been using this as my window manager for work and personal use for about two years now and I have been a very happy user.

I am running into an issue where one of my monitors is not being properly discovered under certain conditions. My desk setup consists of two 4k monitors connected to a single ThinkPad Thunderbolt 3 dock, which connects to my laptop (Thinkpad X1 Carbon gen 6) via a single thunderbolt 3 cable.

Thunderbolt 3 support in Linux has always been a little dicey, but I have been able to get this setup to work by performing the following steps:

1. Log into a tty, no window manager active.
2. Plug in the dock cable. Wait for 10-15 seconds or so, until all monitors are active and mirroring the laptop display (the default behavior in the tty)
3. Start sway
4. Sway starts up, all monitors are detected and running at 60 Hz

I have been using this sequence for a while, and it works well enough for most cases. If you unplug from the dock, everything works as expected. **However,** if I then try to plug the dock back in, sway will only connect to one of my external monitors. `swaymsg -t get_outputs` shows all three outputs, but the non-functioning monitor shows up as "inactive." When I try to explicitly enable the display with `swaymsg output DP-4 enable` or similar, nothing happens.

The result is that, in order to get all monitors working again, I have to kill sway and drop back down to tty, _then_ connect the dock and restart sway. I've been working with this limitation for a while, but I'd like to help investigate and fix the root cause.

I see messages like the following when connecting the dock cable while sway is already running:

```
00:00:19.418 [INFO] [backend/drm/drm.c:811] Modesetting 'DP-4' with '3840x2160@59996mHz'                                                                                                                                              
00:00:19.418 [DEBUG] [backend/drm/drm.c:693] Initializing renderer on connector 'DP-4'                                                                                                                                                
00:00:19.441 [ERROR] [backend/drm/atomic.c:36] DP-4: Atomic commit failed (modeset): Invalid argument                                                                                                                                 
00:00:19.441 [INFO] [backend/drm/drm.c:731] Page-flip failed with primary FB modifiers enabled, retrying without modifiers                                                                                                            
00:00:19.458 [ERROR] [backend/drm/atomic.c:36] DP-4: Atomic commit failed (modeset): Invalid argument                                                                                                                                 
00:00:19.458 [ERROR] [backend/drm/drm.c:744] Failed to initialize renderer on connector 'DP-4': initial page-flip failed                                                                                                                00:00:19.458 [ERROR] [backend/drm/drm.c:817] Failed to initialize renderer for plane                                                                                                                                                  
```
These messages seemed similar to https://github.com/swaywm/sway/issues/5315 (and, it looks like this user was using the same dock as I am); however, I tried re-running with `WLR_DRM_NO_MODIFIERS=1` and did not see any change in behavior. Based on the log messages, it seems like sway is already retrying without modifiers anyways, so I suspect this is a different issue. It seems the issue is probably described more fully in the DRM logs, but I am not really sure on how to begin interpreting those.

Unfortunately, I'm not really clear on where to go from here in the investigation. I have included DRM and sway logs for both the working case (plug dock in prior to starting sway) and non-working case (plug dock in after starting sway). I am happy to file this bug report in upstream projects if the problem lies elsewhere!

To summarize:

I am unable to activate one of my two monitors when connecting a thunderbolt 3 dock to my laptop while sway is running. Both monitors _do_ activate when the dock is connected _prior_ to starting sway. I am trying to investigate why the second monitor cannot activate while sway is running.

Thanks again for any assistance y'all can provide!

**Comment by kennylevinsen** [comment]
>  Requested display configuration exceeds system DDB limitations

So, bandwidth limitation. This can happen if e.g. one display has already been initialized with a more bandwidth-heavy configuration, which `WLR_DRM_NO_MODIFIERS=1` tries to avoid by making *all* outputs run bandwidth conservative.

Note that my experience with this particular series of docks is poor, mostly due to the fact that it uses DisplayPort MST rather than picking a second independent DisplayPort stream off Thunderbolt.

For reference, with my CalDigit dock + extra USB-C to DisplayPort dongle in TB3 passthrough port, my XPS13 with a i7-8550U can drive 2 external 4k@60Hz monitors + its internal 4k@60Hz monitor with no *major* issues. On the other hand, I have used quite a lot of Thinkpad Thunderbolt 3 docks with this laptop as they're popular at various offices, and I can only ever light up 1 external 4k@60Hz monitor in addition to the internal one, or 2x2k@60Hz. Even 2x1440p@60hz leads to one monitor always failing to light up unless I disable the internal panel first.

So, try again with WLR_DRM_NO_MODIFIERS=1 to ensure that everything stays bandwidth conservative, and try to see if disabling the internal panel allows you to enable the dead external output. Also, note that thunderbolt docks on these kinds of laptops behave very differently if connected at *boot* vs. if they're hot-plugged, so keep that in mind during testing.

**Comment by jboyens** [comment]
`WLR_DRM_NO_MODIFIERS=1` fixes the problem for me. I'm not using a dock, however, I'm plugged directly via USB-C -> DisplayPort.

Sway 1.5.1 works just fine with no ENV vars. Sway-HEAD requires this ENV var to drive the second monitor at greater than 1080p. I'm not sure if this is a regression, or merely a better adherence to specifications, but there IS a difference between last released version and the version in `master`

**Comment by emersion** [comment]
Can you check whether reverting https://github.com/swaywm/wlroots/commit/713c1661b742f93a7d2167321837c0d99541ca87 helps?

**Comment by ReeceStevens** [comment]
Thanks all for the speedy response and context about what is going on with DRM!

I retried using `WLR_DRM_NO_MODIFIERS=1` today and, surprisingly, everything started working as expected. I could unplug and re-plug in my monitors and they all would light up at the correct resolution and frame rate. Occasionally only one monitor would light up; however, this also corresponded to my keyboard and mouse not activating, so it seems like something wrong with the activation of some portions of the dock. Unplugging and re-plugging solved the issue.

> Note that my experience with this particular series of docks is poor, mostly due to the fact that it uses DisplayPort MST rather than picking a second independent DisplayPort stream off Thunderbolt.

Boy, it's difficult to find a good quality thunderbolt 3 dock. That's good to know that I'm not alone in this, at least!

@jboyens, RE

> Sway 1.5.1 works just fine with no ENV vars

I am surprised to see this-- I have actually never had this working. I got my second 4k monitor early this year, and it has never worked out of the box; I've always had to do this "drop to tty and restart" trick. I highly suspect my issue is specific to this dock and running both monitors off a single thunderbolt 3 port, and @kennylevinsen's comments seem to confirm this suspicion.

@emersion since I've never had this working, I am assuming your request to revert that wlroots commit was directed towards @jboyens and not myself

**Comment by jboyens** [comment]
@emersion I reverted wlroots back to https://github.com/swaywm/wlroots/commit/154fe8696f40151dae999cfc69299aef5132dd14 and it did indeed fix the issue for me. So, while detection is slow, it is necessary. I did also try the command in the commit message and it too, solved the issue.

**Comment by jboyens** [comment]
N.B. I did reboot between each test to make sure a previous detection wasn't held over to ruin a test.

**Comment by emersion** [comment]
OK. I'm going to close this as a duplicate of https://github.com/swaywm/wlroots/issues/1877.

@jboyens You're experiencing https://github.com/swaywm/wlroots/issues/2499 then. I'm sending a revert for that commit now.
